website/docs/usage/parameters.md
---
id: parameters
title: Parameters
---

`Parameters` is an object type to specify detailed checker behavior.
It has many parameters.
But, few parameters are useful to users, and the others are not needed to be set manually.

You can look at the whole type definition [here](https://github.com/makenowjust-labs/recheck/blob/main/packages/recheck/index.d.ts) (TypeScript) or [here](https://github.com/makenowjust-labs/recheck/blob/main/modules/recheck-common/shared/src/main/scala/codes/quine/labs/recheck/common/Parameters.scala) (Scala).

## `accelerationMode`

- Type: `'auto' | 'on' | 'off'` (TypeScript), `AccelerationMode` (Scala)
- Default: `'auto'` (TypeScript), `AccelerationMode.Auto` (Scala)

The type of acceleration mode strategy on fuzzing.

There are three acceleration mode strategies.

- `auto` uses acceleration mode as default. However, if the regular expression has backreferences, it turns off the acceleration mode.
- `on` turns on the acceleration mode.
- `off` turns off the acceleration mode.

## `attackLimit`

- Type: `number` (TypeScript), `Int` (Scala)
- Default: `1500000000`

The upper limit on the number of characters read by VM on the attack.

## `attackTimeout`

- Type: `number | null` (TypeScript), `Duration` (Scala)
- Default: `1000` (TypeScript), `Duration(1, SECONDS)` (Scala)

The upper limit of matching time on the attack.

## `checker`

- Type: `'auto' | 'fuzz' | 'automaton'` (TypeScript), `Checker` (Scala)
- Default: `'auto'` (TypeScript), `Checker.Auto` (Scala)

The type of checker to be used.

There are three checker types.

- `auto` checker uses the criteria to decide which algorithm is better to use against a regular expression, the algorithm based on automata theory or the fuzzing algorithm.
- `fuzz` checker uses the fuzzing algorithm with static analysis.
- `automaton` checker uses the algorithm based on automata theory.

## `crossoverSize`

- Type: `number` (TypeScript), `Int` (Scala)
- Default: `25`

The number of crossover on each generation.

## `heatRatio`

- Type: `number` (TypeScript), `Double` (Scala)
- Default: `0.001`

The ratio of the number of characters read to the maximum number to be considered as a hot spot.

## `incubationLimit`

- Type: `number` (TypeScript), `Int` (Scala)
- Default: `25000`

The upper limit on the number of characters read by VM on incubation.

## `incubationTimeout`

- Type: `number | null` (TypeScript), `Duration` (Scala)
- Default: `250` (TypeScript), `Duration(250, MILLISECONDS)` (Scala)

The upper limit of matching time on incubation.

## `logger`

- Type: `(message: string) => void` (TypeScript), `Option[Context.Logger]` (Scala)
- Default: `null` (TypeScript), `None` (Scala)

The logger function to record execution traces.

To disable the logging, `null` in TypeScript or `None` in Scala should be passed.

## `maxAttackStringSize`

- Type: `number` (TypeScript), `Int` (Scala)
- Default: `300000`

The maximum length of the attack string on fuzzing.

## `maxDegree`

- Type: `number` (TypeScript), `Int` (Scala)
- Default: `4`

The maximum degree to be considered in fuzzing.

## `maxGeneStringSize`

- Type: `number` (TypeScript), `Int` (Scala)
- Default: `2400`

The maximum length of the gene string on fuzzing.

## `maxGenerationSize`

- Type: `number` (TypeScript), `Int` (Scala)
- Default: `100`

The maximum size of each generation on fuzzing.

## `maxInitialGenerationSize`

- Type: `number` (TypeScript), `Int` (Scala)
- Default: `500`

The maximum size of the initial generation on fuzzing.

## `maxIteration`

- Type: `number` (TypeScript), `Int` (Scala)
- Default: `10`

The maximum number of fuzzing iteration.

## `maxNFASize`

- Type: `number` (TypeScript), `Int` (Scala)
- Default: `35000`

The maximum size of NFA to determine which algorithm is used.

## `maxPatternSize`

- Type: `number` (TypeScript), `Int` (Scala)
- Default: `1500`

The maximum size of the regular expression pattern to determine which algorithm is used.

## `maxRecallStringSize`

- Type: `number` (TypeScript), `Int` (Scala)
- Default: `300000`

The maximum length of the attack string on recall validation.

## `maxRepeatCount`

- Type: `number` (TypeScript), `Int` (Scala)
- Default: `30`

The maximum number of sum of repetition quantifier’s repeat counts to determine which algorithm is used.

## `maxSimpleRepeatCount`

- Type: `number` (TypeScript), `Int` (Scala)
- Default: `30`

The maximum number of each repetition quantifier’s repeat count on `static` seeding.

## `mutationSize`

- Type: `number` (TypeScript), `Int` (Scala)
- Default: `50`

The number of mutation on each generation.

## `randomSeed`

- Type: `number` (TypeScript), `Long` (Scala)
- Default: `0`

The PRNG seed number.

## `recallLimit`

- Type: `number` (TypeScript), `Int` (Scala)
- Default: `1500000000`

The upper limit on the number of characters read by VM on the recall validation.

## `recallTimeout`

- Type: `number | null` (TypeScript), `Duration` (Scala)
- Default: `-1000` (TypeScript), `Duration(-1, SECONDS)` (Scala)

The upper limit of matching time on the recall validation.

If this value is negative, then the recall validation is skipped.

## `seeder`

- Type: `'static' | 'dynamic'` (TypeScript), `Seeder` (Scala)
- Default: `'static'` (TypeScript), `Seeder.Static` (Scala)

The type of seeder to be used in fuzzing.

There are two seeders.

- `static` seeder uses the seeding algorithm based on the automata theory.
- `dynamic` seeder uses the seeding algorithm with dynamic analysis.

## `seedingLimit`

- Type: `number` (TypeScript), `Int` (Scala)
- Default: `1000`

The upper limit on the number of characters read by VM on `dynamic` seeding.

## `seedingTimeout`

- Type: `number | null` (TypeScript), `Duration` (Scala)
- Default: `100` (TypeScript), `Duration(100, MILLISECONDS)` (Scala)

The upper limit of matching time on `dynamic` seeding.

## `timeout`

- Type: `number | null` (TypeScript), `Duration` (Scala)
- Default: `10000` (TypeScript), `Duration(10, SECONDS)` (Scala)

The upper limit of checking time.

If the checking time exceeds this limit, the result will be reported as `timeout`. If the value is positive infinite in Scala or `null` in TypeScript, the result never becomes `timeout`.

The `timeout` time begins to be measured as soon as the check starts. Note that the `timeout` does not occur while the input is in the queue waiting to be checked.

In TypeScript, a number value is treated as in milliseconds.


CHANGELOG.md
# dev

Changes:

- [#357](https://github.com/makenowjust-labs/recheck/pull/357) Add `Default` prefix to default parameter values ([@makenowjust](https://github.com/makenowjust))
- [#358](https://github.com/makenowjust-labs/recheck/pull/358) Sort `Parameters` names in lexicographic order ([@makenowjust](https://github.com/makenowjust))

Fixes:

- [#341](https://github.com/makenowjust-labs/recheck/pull/341) Fix zero-width assertion in `automaton` checker ([@makenowjust](https://github.com/makenowjust))

Misc:

- [#337](https://github.com/makenowjust-labs/recheck/pull/337) Increase timeout on `main.ts` test ([@makenowjust](https://github.com/makenowjust))
- [#343](https://github.com/makenowjust-labs/recheck/pull/343) Enable Algolia DocSearch ([@makenowjust](https://github.com/makenowjust))
- [#360](https://github.com/makenowjust-labs/recheck/pull/360) Enable log on playground ([@makenowjust](https://github.com/makenowjust))

# 4.3.0 (2022-02-05)

New Features:

- [#289](https://github.com/makenowjust-labs/recheck/pull/289) Add `worker` backend to the NPM package ([@makenowjust](https://github.com/makenowjust))

Changes:

- [#283](https://github.com/makenowjust-labs/recheck/pull/283) Rename `hybrid` to `auto` ([@makenowjust](https://github.com/makenowjust))
- [#332](https://github.com/makenowjust-labs/recheck/pull/332) Rename `labo` to `labs` ([@makenowjust](https://github.com/makenowjust))

Fixes:

- [#284](https://github.com/makenowjust-labs/recheck/pull/284) Add missing field `AttackPattern#pattern` in TypeScript definition ([@makenowjust](https://github.com/makenowjust))

Misc:

- [#279](https://github.com/makenowjust-labs/recheck/pull/279) Update `CHANGELOG.md` automatically by GitHub Actions ([@makenowjust](https://github.com/makenowjust))
- [#280](https://github.com/makenowjust-labs/recheck/pull/280) Handle SBT deps by renovate correctly ([@makenowjust](https://github.com/makenowjust))
- [#281](https://github.com/makenowjust-labs/recheck/pull/281) Update scala-library, scala-reflect to 2.13.8 ([@makenowjust](https://github.com/makenowjust))
- [#282](https://github.com/makenowjust-labs/recheck/pull/282) Add note section to ESLint plugin document ([@makenowjust](https://github.com/makenowjust))
- [#286](https://github.com/makenowjust-labs/recheck/pull/286) Update `CHANGELOG.md` format ([@makenowjust](https://github.com/makenowjust))
- [#295](https://github.com/makenowjust-labs/recheck/pull/295) Renew document website powered by Docusaurus ([@makenowjust](https://github.com/makenowjust))
- [#313](https://github.com/makenowjust-labs/recheck/pull/313) Remove `mdoc` to build the document ([@makenowjust](https://github.com/makenowjust))
- [#314](https://github.com/makenowjust-labs/recheck/pull/314) Don't ignore to build website on CI test ([@makenowjust](https://github.com/makenowjust))
- [#318](https://github.com/makenowjust-labs/recheck/pull/318) Fix logo style ([@makenowjust](https://github.com/makenowjust))
- [#320](https://github.com/makenowjust-labs/recheck/pull/320) Improve documentation ([@makenowjust](https://github.com/makenowjust))

# 4.2.2 (2022-01-14)

Fixes:

- Fix `handleLine` in `Agent` for large data ([#266](https://github.com/makenowjust-labs/recheck/pull/266))
- Prefer to use `native` backend on `RECHECK_BACKEND=auto` ([#268](https://github.com/makenowjust-labs/recheck/pull/268))

# 4.2.1 (2022-01-08)

Fixes:

- Don't bundle recheck in eslint-plugin-redos ([#260](https://github.com/makenowjust-labs/recheck/pull/260))
- Use `require` directly instead of `module.require` ([#261](https://github.com/makenowjust-labs/recheck/pull/261))

# 4.2.0 (2022-01-08)

Changes:

- Split `common` package into another module `recheck-common` ([#202](https://github.com/makenowjust-labs/recheck/pull/202))
- Increase default parameter values ([#202](https://github.com/makenowjust-labs/recheck/pull/202))
- Restrict repeats contained in gene ([#202](https://github.com/makenowjust-labs/recheck/pull/202))
- Add `Parameters` as immutable version of `Config` ([#202](https://github.com/makenowjust-labs/recheck/pull/202))
- Add `accelerationMode` parameter to specify acceleration mode behavior ([#211](https://github.com/makenowjust-labs/recheck/pull/211))
- Rename default parameter constant names ([#211](https://github.com/makenowjust-labs/recheck/pull/211))
- Add the new default seeder named `StaticSeeder` ([#227](https://github.com/makenowjust-labs/recheck/pull/227))
- Decrease some parameters default values ([#227](https://github.com/makenowjust-labs/recheck/pull/227))
- Improve repetition count estimation ([#227](https://github.com/makenowjust-labs/recheck/pull/227))
- Use JavaScript style representation for `AttackPattern#toString` ([#229](https://github.com/makenowjust-labs/recheck/pull/229))
- Apply look-ahead optimization to VM ([#233](https://github.com/makenowjust-labs/recheck/pull/233))
- Add recall validation ([#234](https://github.com/makenowjust-labs/recheck/pull/234))
- Simplify automaton without input terminator distinction ([#240](https://github.com/makenowjust-labs/recheck/pull/240))
- Add `recheck-jar` package ([#248](https://github.com/makenowjust-labs/recheck/pull/248))
- Improve `StaticSeeding.simplify` against look-around near by repeat ([#255](https://github.com/makenowjust-labs/recheck/pull/255))

Fixes:

- Fix semaphore usage in test ([#230](https://github.com/makenowjust-labs/recheck/pull/230))
- Fix `sbt` file issues ([#231](https://github.com/makenowjust-labs/recheck/pull/231))
- Check integer overflow on matching steps ([#232](https://github.com/makenowjust-labs/recheck/pull/232))
- Fix loop detection in `StaticSeeder` ([#235](https://github.com/makenowjust-labs/recheck/pull/235))
- Returns iterator in `FuzzChecker`  ([#236](https://github.com/makenowjust-labs/recheck/pull/236))
- Rewrite `recheck` package implementation ([#248](https://github.com/makenowjust-labs/recheck/pull/248))

modules/recheck-common/shared/src/main/scala/codes/quine/labs/recheck/common/Parameters.scala
package codes.quine.labs.recheck.common

import scala.concurrent.duration._

/** Parameters is an immutable representation of parameters.
  *
  * @param accelerationMode
  *
  * The type of acceleration mode strategy on fuzzing.
  *
  * There are three acceleration mode strategies.
  *
  *   - `auto` uses acceleration mode as default. However, if the regular expression has backreferences, it turns off
  *     the acceleration mode.
  *   - `on` turns on the acceleration mode.
  *   - `off` turns off the acceleration mode.
  *
  * (default: `AccelerationMode.Auto`)
  *
  * @param attackLimit
  *
  * The upper limit on the number of characters read by VM on the attack.
  *
  * (default: `1500000000`)
  *
  * @param attackTimeout
  *
  * The upper limit of matching time on the attack.
  *
  * (default: `Duration(1, SECONDS)`)
  *
  * @param checker
  *
  * The type of checker to be used.
  *
  * There are three checker types.
  *
  *   - `auto` checker uses the criteria to decide which algorithm is better to use against a regular expression, the
  *     algorithm based on automata theory or the fuzzing algorithm.
  *   - `fuzz` checker uses the fuzzing algorithm with static analysis.
  *   - `automaton` checker uses the algorithm based on automata theory.
  *
  * (default: `Checker.Auto`)
  *
  * @param crossoverSize
  *
  * The number of crossover on each generation.
  *
  * (default: `25`)
  *
  * @param heatRatio
  *
  * The ratio of the number of characters read to the maximum number to be considered as a hot spot.
  *
  * (default: `0.001`)
  *
  * @param incubationLimit
  *
  * The upper limit on the number of characters read by VM on incubation.
  *
  * (default: `25000`)
  *
  * @param incubationTimeout
  *
  * The upper limit of matching time on incubation.
  *
  * (default: `Duration(250, MILLISECONDS)`)
  *
  * @param logger
  *
  * The logger function to record execution traces.
  *
  * To disable the logging, `null` in TypeScript or `None` in Scala should be passed.
  *
  * (default: `None`)
  *
  * @param maxAttackStringSize
  *
  * The maximum length of the attack string on fuzzing.
  *
  * (default: `300000`)
  *
  * @param maxDegree
  *
  * The maximum degree to be considered in fuzzing.
  *
  * (default: `4`)
  *
  * @param maxGeneStringSize
  *
  * The maximum length of the gene string on fuzzing.
  *
  * (default: `2400`)
  *
  * @param maxGenerationSize
  *
  * The maximum size of each generation on fuzzing.
  *
  * (default: `100`)
  *
  * @param maxInitialGenerationSize
  *
  * The maximum size of the initial generation on fuzzing.
  *
  * (default: `500`)
  *
  * @param maxIteration
  *
  * The maximum number of fuzzing iteration.
  *
  * (default: `10`)
  *
  * @param maxNFASize
  *
  * The maximum size of NFA to determine which algorithm is used.
  *
  * (default: `35000`)
  *
  * @param maxPatternSize
  *
  * The maximum size of the regular expression pattern to determine which algorithm is used.
  *
  * (default: `1500`)
  *
  * @param maxRecallStringSize
  *
  * The maximum length of the attack string on recall validation.
  *
  * (default: `300000`)
  *
  * @param maxRepeatCount
  *
  * The maximum number of sum of repetition quantifier’s repeat counts to determine which algorithm is used.
  *
  * (default: `30`)
  *
  * @param maxSimpleRepeatCount
  *
  * The maximum number of each repetition quantifier’s repeat count on `static` seeding.
  *
  * (default: `30`)
  *
  * @param mutationSize
  *
  * The number of mutation on each generation.
  *
  * (default: `50`)
  *
  * @param randomSeed
  *
  * The PRNG seed number.
  *
  * (default: `0`)
  *
  * @param recallLimit
  *
  * The upper limit on the number of characters read by VM on the recall validation.
  *
  * (default: `1500000000`)
  *
  * @param recallTimeout
  *
  * The upper limit of matching time on the recall validation.
  *
  * If this value is negative, then the recall validation is skipped.
  *
  * (default: `Duration(-1, SECONDS)`)
  *
  * @param seeder
  *
  * The type of seeder to be used in fuzzing.
  *
  * There are two seeders.
  *
  *   - `static` seeder uses the seeding algorithm based on the automata theory.
  *   - `dynamic` seeder uses the seeding algorithm with dynamic analysis.
  *
  * (default: `Seeder.Static`)
  *
  * @param seedingLimit
  *
  * The upper limit on the number of characters read by VM on `dynamic` seeding.
  *
  * (default: `1000`)
  *
  * @param seedingTimeout
  *
  * The upper limit of matching time on `dynamic` seeding.
  *
  * (default: `Duration(100, MILLISECONDS)`)
  *
  * @param timeout
  *
  * The upper limit of checking time.
  *
  * If the checking time exceeds this limit, the result will be reported as `timeout`. If the value is positive infinite
  * in Scala or `null` in TypeScript, the result never becomes `timeout`.
  *
  * The `timeout` time begins to be measured as soon as the check starts. Note that the `timeout` does not occur while
  * the input is in the queue waiting to be checked.
  *
  * In TypeScript, a number value is treated as in milliseconds.
  *
  * (default: `Duration(10, SECONDS)`)
  */
final case class Parameters(
    accelerationMode: AccelerationMode = Parameters.DefaultAccelerationMode,
    attackLimit: Int = Parameters.DefaultAttackLimit,
    attackTimeout: Duration = Parameters.DefaultAttackTimeout,
    checker: Checker = Parameters.DefaultChecker,
    crossoverSize: Int = Parameters.DefaultCrossoverSize,
    heatRatio: Double = Parameters.DefaultHeatRatio,
    incubationLimit: Int = Parameters.DefaultIncubationLimit,
    incubationTimeout: Duration = Parameters.DefaultIncubationTimeout,
    logger: Option[Context.Logger] = Parameters.DefaultLogger,
    maxAttackStringSize: Int = Parameters.DefaultMaxAttackStringSize,
    maxDegree: Int = Parameters.DefaultMaxDegree,
    maxGeneStringSize: Int = Parameters.DefaultMaxGeneStringSize,
    maxGenerationSize: Int = Parameters.DefaultMaxGenerationSize,
    maxInitialGenerationSize: Int = Parameters.DefaultMaxInitialGenerationSize,
    maxIteration: Int = Parameters.DefaultMaxIteration,
    maxNFASize: Int = Parameters.DefaultMaxNFASize,
    maxPatternSize: Int = Parameters.DefaultMaxPatternSize,
    maxRecallStringSize: Int = Parameters.DefaultMaxRecallStringSize,
    maxRepeatCount: Int = Parameters.DefaultMaxRepeatCount,
    maxSimpleRepeatCount: Int = Parameters.DefaultMaxSimpleRepeatCount,
    mutationSize: Int = Parameters.DefaultMutationSize,
    randomSeed: Long = Parameters.DefaultRandomSeed,
    recallLimit: Int = Parameters.DefaultRecallLimit,
    recallTimeout: Duration = Parameters.DefaultRecallTimeout,
    seeder: Seeder = Parameters.DefaultSeeder,
    seedingLimit: Int = Parameters.DefaultSeedingLimit,
    seedingTimeout: Duration = Parameters.DefaultSeedingTimeout,
    timeout: Duration = Parameters.DefaultTimeout
)

object Parameters {
  // $COVERAGE-OFF$

  /** The default value of [[Parameters.accelerationMode]]. */
  val DefaultAccelerationMode: AccelerationMode = AccelerationMode.Auto

  /** The default value of [[Parameters.attackLimit]]. */
  val DefaultAttackLimit: Int = 1500000000

  /** The default value of [[Parameters.attackTimeout]]. */
  val DefaultAttackTimeout: Duration = Duration(1, SECONDS)

  /** The default value of [[Parameters.checker]]. */
  val DefaultChecker: Checker = Checker.Auto

  /** The default value of [[Parameters.crossoverSize]]. */
  val DefaultCrossoverSize: Int = 25

  /** The default value of [[Parameters.heatRatio]]. */
  val DefaultHeatRatio: Double = 0.001

  /** The default value of [[Parameters.incubationLimit]]. */
  val DefaultIncubationLimit: Int = 25000

  /** The default value of [[Parameters.incubationTimeout]]. */
  val DefaultIncubationTimeout: Duration = Duration(250, MILLISECONDS)

  /** The default value of [[Parameters.logger]]. */
  val DefaultLogger: Option[Context.Logger] = None

  /** The default value of [[Parameters.maxAttackStringSize]]. */
  val DefaultMaxAttackStringSize: Int = 300000

  /** The default value of [[Parameters.maxDegree]]. */
  val DefaultMaxDegree: Int = 4

  /** The default value of [[Parameters.maxGeneStringSize]]. */
  val DefaultMaxGeneStringSize: Int = 2400

  /** The default value of [[Parameters.maxGenerationSize]]. */
  val DefaultMaxGenerationSize: Int = 100

  /** The default value of [[Parameters.maxInitialGenerationSize]]. */
  val DefaultMaxInitialGenerationSize: Int = 500

  /** The default value of [[Parameters.maxIteration]]. */
  val DefaultMaxIteration: Int = 10

  /** The default value of [[Parameters.maxNFASize]]. */
  val DefaultMaxNFASize: Int = 35000

  /** The default value of [[Parameters.maxPatternSize]]. */
  val DefaultMaxPatternSize: Int = 1500

  /** The default value of [[Parameters.maxRecallStringSize]]. */
  val DefaultMaxRecallStringSize: Int = 300000

  /** The default value of [[Parameters.maxRepeatCount]]. */
  val DefaultMaxRepeatCount: Int = 30

  /** The default value of [[Parameters.maxSimpleRepeatCount]]. */
  val DefaultMaxSimpleRepeatCount: Int = 30

  /** The default value of [[Parameters.mutationSize]]. */
  val DefaultMutationSize: Int = 50

  /** The default value of [[Parameters.randomSeed]]. */
  val DefaultRandomSeed: Long = 0

  /** The default value of [[Parameters.recallLimit]]. */
  val DefaultRecallLimit: Int = 1500000000

  /** The default value of [[Parameters.recallTimeout]]. */
  val DefaultRecallTimeout: Duration = Duration(-1, SECONDS)

  /** The default value of [[Parameters.seeder]]. */
  val DefaultSeeder: Seeder = Seeder.Static

  /** The default value of [[Parameters.seedingLimit]]. */
  val DefaultSeedingLimit: Int = 1000

  /** The default value of [[Parameters.seedingTimeout]]. */
  val DefaultSeedingTimeout: Duration = Duration(100, MILLISECONDS)

  /** The default value of [[Parameters.timeout]]. */
  val DefaultTimeout: Duration = Duration(10, SECONDS)

  // $COVERAGE-ON$
}
